### ISSUE TITLE
MBXMapKit Example App high CPU usage

### ISSUE THREAD

**Issue by ndm-io** [mapbox/mbxmapkit#163]
I implemented an offline map database in my own app and noticed that whenever offline map tiles were added CPU usage would go up over 175% and would not idle once the map was rendered.

I had a look through the code for MBXRasterTileRenderer and noticed that in -canDrawMapRect: zoomScale: the call to  self -setNeedsDisplayInMapRect: zoomScale: in the result block occurs every time.

Moving the call to within the if block seems to have solved the problem, hopefully without new and unintended consequences!

```c
if ( ! tileActive) {
            [(MKTileOverlay *)weakSelf.overlay loadTileAtPath:path result:^(NSData *tileData, NSError *error) {
                @synchronized(weakSelf) {
                    [weakSelf.activeDownloads removeObject:xyz];
                }
                if (tileData) {
                    NSData *tileDataCopy = [[NSData alloc] initWithBytes:tileData.bytes length:tileData.length];
                    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                        CGDataProviderRef provider = CGDataProviderCreateWithCFData((CFDataRef)tileDataCopy);
                        CGImageRef imageRef = nil;
                        if ([[weakSelf class] dataIsPNG:tileDataCopy]) {
                            imageRef = CGImageCreateWithPNGDataProvider(provider, nil, NO, kCGRenderingIntentDefault);
                        } else if ([[weakSelf class] dataIsJPEG:tileDataCopy]) {
                            imageRef = CGImageCreateWithJPEGDataProvider(provider, nil, NO, kCGRenderingIntentDefault);
                        }
                        if (imageRef) {
                            @synchronized(weakSelf) {
                                [[weakSelf class] addImageData:tileDataCopy
                                                    toRenderer:weakSelf
                                                        forXYZ:xyz
                                                 usingBigTiles:usingBigTiles];
                            }
                        }
                        CGImageRelease(imageRef);
                        CGDataProviderRelease(provider);
                    });
                    
                    //Moving this line here has prevented constant redrawing of the map
                    [weakSelf setNeedsDisplayInMapRect:mapRect zoomScale:zoomScale];
                }
                //Originally the call was here
            }];
        } else {
            [weakSelf setNeedsDisplayInMapRect:mapRect zoomScale:zoomScale];
        }

**Comment by meierjo** [comment]
check out #153.  The fix there helped my high CPU as well as a random crash ...

**Comment by s219** [comment]
Now that I see this issue and #153 discussed together, I am betting they are one in the same.  Instead of the fix proposed above, which could potentially lead to missing tiles from an online map -- since the ``setNeedsDisplayInMapRect`` call would have the (presumably) desirable effect of re-queuing an online tile for download if the data was bad -- I'd suggest leaving the ``setNeedsDisplayInMapRect`` call in the original location, but use an ``if`` to skip it if ``loadTileAtPath`` returns an error associated with missing offline tiles:

```
if ([error code] != MBXMapKitErrorCodeOfflineMapHasNoDataForURL) [weakSelf setNeedsDisplayInMapRect:mapRect zoomScale:zoomScale];
```

So ``setNeedsDisplayInMapRect`` will be called in all cases except the case of a missing offline tile.

That appears to block repeated calls to load an offline tile that does not exist, and avoids high CPU usage and crashes were were discussing on #153 .  It also eliminates the need for that blank tile hack in #153 , so it allows over/under zooming of tiles at other zoom levels.  Essentially, if there is no available offline tile for a mapRect, any previous rendering is left in place and we do not attempt to load the tile any further.

**Comment by meierjo** [comment]
@s219 I'll have some time to test your comment out in the next few days - sounds like the best way to proceed !

**Comment by s219** [comment]
Well, I saw a crash today with online maps, which originated from the ``setNeedsDisplayInMapRect`` call and veered off into the #153 crash.  Need to study the code again -- may want to revisit and think about scenarios when that call needs to be skipped for online maps -- might be other error conditions, or cases where tileData is bad.  @ndm-io may have had a safe approach up top of the thread.

**Comment by s219** [comment]
I did some stress testing with online maps, and was able to create an endless looping condition similar to what we saw in #153 for offline maps when using my fix suggested up above.  It took some doing and may not be a common scenario, but I showed it could happen.  Based on that, I think @ndm-io has the best solution (whether offline or online), by calling ``setNeedsDisplayInMapRect`` inside the ``if(tileData)`` check.  May even be that we should move that call inside the ``if(imageRef)`` check too. 

I imagine the MapBox guys are busy with WWDC this week, but would be good to get some feedback here.  It's possible that the location of the ``setNeedsDisplayInMapRect`` call in 0.7 is in fact a bug, and it should be up inside the ``tileData`` check as @ndm-io suggested.  Of if it was meant to be where it is, I think we really need some sort of limit so that the whole ``setNeedsDisplayInMapRect `` --> ``canDrawMapRect`` cycle doesn't go into an endless loop in certain scenarios.
